modularity_maximization/modularity.py

```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+import networkx as nx
+from scipy import sparse
+from scipy.linalg import eig
+from itertools import product
+
+
+def transform_net_and_part(network,partition):
+    '''
+    Accepts an input network and a community partition (keys are nodes,
+    values are community ID) and returns a version of the network and
+    partition with nodes in the range 0,...,len(G.nodes())-1.  This
+    lets you directly map edges to elements of the modularity matrix.
+
+    Returns the modified network and partition.
+    '''
+    network = nx.convert_node_labels_to_integers(network, first_label=0, label_attribute="node_name")
+    node_to_name = nx.get_node_attributes(network, 'node_name')
+    # reverse the node_name dict to flip the partition
+    name_to_node = {v:k for k,v in node_to_name.items()}
+    int_partition = {}
+    for k in partition:
+        int_partition[name_to_node[k]] = partition[k]
+    return network,int_partition
+
+
+def reverse_partition(partition):
+    '''
+    Accepts an input graph partition in the form node:community_id and returns
+    a dictionary of the form community_id:[node_1,node_2,...].
+    '''
+    reverse_partition = {}
+    for p in partition:
+        if partition[p] in reverse_partition:
+            reverse_partition[partition[p]].append(p)
+        else:
+            reverse_partition[partition[p]] = [p]
+    return reverse_partition
+
+
+def modularity(network, partition):
+    '''
+    Computes the modularity; works for Directed and Undirected Graphs, both
+    unweighted and weighted.
+    '''
+    # put the network and partition into integer node format
+    network,partition = transform_net_and_part(network,partition)
+    # get the modularity matrix
+    Q = get_base_modularity_matrix(network)
+    if type(network) == nx.Graph:
```

Comment:
here could be further trimmed as:
```python
if nx.is_weighted(network): 
    norm_fac = nx.to_scipy_sparse_matrix(network).sum()
elif type(network) == nx.Graph:
    norm_fac = 2.*(network.number_of_edges())
elif type(network) == nx.DiGraph:
    norm_fac = 1.*network.number_of_edges()
else:
    print('Invalid graph type')
    raise TypeError
```

Reply:
Thanks for the reply!  I'll try to get this in there ASAP.